### PR TITLE
implement sign in experience

### LINKED ISSUE #12: Implement complete Sign In experience
This issue relies on following issues to be completed first:

- https://github.com/mudrila/fin-plan/issues/5
- https://github.com/mudrila/fin-plan/issues/6
- https://github.com/mudrila/fin-plan/issues/7
- https://github.com/mudrila/fin-plan/issues/2
- https://github.com/mudrila/fin-plan/issues/8

The expected deliverables are:

- When user submits "Sign In" form - the form goes into "loading" state. All the form fields are getting into `disabled` state, as well as Submit button. The submit button itself also changes it's text from "Submit" to "Submitting..." - as long as form processing is still going.
- On the server side - we should verify that user with such email exists. If that's not the case - we should be returning `JSON` with following content: `{ errorMessage: "User with given email / password combination was not found" }`
- Also, we need to verify that provided `password` matches password hash stored in the DB. If that's not the case - we should be returning very same error: `{ errorMessage: "User with given email / password combination was not found" }`
- When client-side receives response with `errorMessage` field - it should display error toast with returned `errorMessage`
- For the error toast - we should be using `sonner` library: https://sonner.emilkowal.ski/
- If form submissions succeeded - user should go into "authed" state and being redirected to the application's home page under `/app` path
- When user is "authed" - we should redirect him from `/sign-in` page and `/sign-up` page to homepage under `/app` path
- When user is not "authed" - we should be redirecting him from all `/app/*` paths to `/sign-in`

### PR DESCRIPTION
**Pull request by Artemh59**
Closes #12

### CODE REVIEW

File: src/app/components/organisms/Forms/SignInForm/SignInForm.tsx [hunk 1]
```diff
@@ -12,14 +14,34 @@ export default function SignInForm() {
...
+  const [isPending, startTransition] = useTransition();
 
   const isFormValid = !emailError && !passwordError;
 
   const handleSubmit = (event: React.FormEvent) => {
     event.preventDefault();
 
     if (isFormValid) {
-      alert('Sign in');
+      startTransition(async () => {
```
Comment (mudrila) [inline src/app/components/organisms/Forms/SignInForm/SignInForm.tsx:25]: Same as for `signOut` - there's specific `signIn` function, that is imported from the same place.

Then the usage will look like this (adjust accordingly to be compatible with `startTransition` or use code snippet I've shared - whatever you'll prefer, sir)

```
try {
        const result = await signIn('email-and-password', {
          redirectTo: '/app',
          email,
          password,
          redirect: false,
        });
        if (!result || !!result.error || !result.ok) {
          toast.error('User with given email and password was not found', {
            duration: 3000,
          });
        } else if (result.ok) {
          toast.success('Welcome on board!', { duration: 3000 });
          trackVercel('Login Success', { email });
          window.location.href = `/app?t=${Date.now()}`;
        }
      } catch (e) {
        logError(e, 'Unexpected error while logging in');
        toast.error('Unexpected error while processing login. Please, try again later', {
          duration: 3000,
        });
      } finally {
        setLoading(false);
      }
```

The actual sign in handling will be happening inside these lines of code: https://github.com/mudrila/fin-plan/blob/main/src/utils/auth.ts#L17-L61

File: src/app/api/auth/sign-in/route.ts
Comment (mudrila) [inline src/app/api/auth/sign-in/route.ts:1]: And we don't need this file

**Review by mudrila — CHANGES_REQUESTED** [review]
Sign in should be happening differently sir

File: src/app/components/organisms/Forms/SignInForm/SignInForm.tsx (hunk 1 above)
Comment (Artemh59) [reply src/app/components/organisms/Forms/SignInForm/SignInForm.tsx:25]: Sorry, where can I import it?(trackVercel, logError)

Comment (mudrila) [reply src/app/components/organisms/Forms/SignInForm/SignInForm.tsx:25]: Ignore `trackVercel`
Replace `logError` with `console.error`

Apologies for the confusion, too quick copy-pasting 😄

Comment (Artemh59) [reply src/app/components/organisms/Forms/SignInForm/SignInForm.tsx:25]: hah, okey, thank u

**Review by mudrila — APPROVED** [review]
Yes - exactly what I was envisioning
@Artemh59 Just to confirm - within this PR - all the user flows regarding auth are working? User can sign up, sign out and sign in again - all the errors are handled properly and no unexpected errors, is this right?

**Comment by Artemh59** [comment]
yes, but we have some 'problems' with password strong check and password check. Where we check the password and it contains '.', use can't go in, but password's strong check is full. I think it must be same, but I dont know what way is correct

**Comment by mudrila** [comment]
> yes, but we have some 'problems' with password strong check and password check. Where we check the password and it contains '.', use can't go in, but password's strong check is full. I think it must be same, but I dont know what way is correct

So we can't have `.` symbol in the password? 
Yeah, that's just minor bug - I'll put a card for it and we will tackle it later.

**Comment by Artemh59** [comment]
Снимок экрана 2024-09-26 в 23 27 53

**Pull request merged**